crypto/dh/dh_rfc7919.c

```diff
@@ -47,28 +77,48 @@ DH *DH_new_by_nid(int nid)
 int DH_get_nid(const DH *dh)
 {
     int nid;
+    const BIGNUM *p, *q, *g;
+    BIGNUM *qc = NULL;
+
+    FFC_PARAMS_get0_pqg(&dh->params, &p, &q, &g);
 
-    if (BN_get_word(dh->g) != 2)
+    if (BN_get_word(g) != 2)
         return NID_undef;
-    if (!BN_cmp(dh->p, &_bignum_ffdhe2048_p))
+
+    if (!BN_cmp(p, &_bignum_ffdhe2048_p))
         nid = NID_ffdhe2048;
-    else if (!BN_cmp(dh->p, &_bignum_ffdhe3072_p))
+    else if (!BN_cmp(p, &_bignum_ffdhe3072_p))
         nid = NID_ffdhe3072;
-    else if (!BN_cmp(dh->p, &_bignum_ffdhe4096_p))
+    else if (!BN_cmp(p, &_bignum_ffdhe4096_p))
         nid = NID_ffdhe4096;
-    else if (!BN_cmp(dh->p, &_bignum_ffdhe6144_p))
+    else if (!BN_cmp(p, &_bignum_ffdhe6144_p))
         nid = NID_ffdhe6144;
-    else if (!BN_cmp(dh->p, &_bignum_ffdhe8192_p))
+    else if (!BN_cmp(p, &_bignum_ffdhe8192_p))
         nid = NID_ffdhe8192;
+#ifndef FIPS_MODE
+    else if (!BN_cmp(p, &_bignum_modp_1536_p))
+        nid = NID_modp_1536;
+#endif
+    else if (!BN_cmp(p, &_bignum_modp_2048_p))
+        nid = NID_modp_2048;
+    else if (!BN_cmp(p, &_bignum_modp_3072_p))
+        nid = NID_modp_3072;
+    else if (!BN_cmp(p, &_bignum_modp_4096_p))
+        nid = NID_modp_4096;
+    else if (!BN_cmp(p, &_bignum_modp_6144_p))
+        nid = NID_modp_6144;
+    else if (!BN_cmp(p, &_bignum_modp_8192_p))
+        nid = NID_modp_8192;
     else
         return NID_undef;
-    if (dh->q != NULL) {
-        BIGNUM *q = BN_dup(dh->p);
 
-        /* Check q = p * 2 + 1 we already know q is odd, so just shift right */
-        if (q == NULL || !BN_rshift1(q, q) || !BN_cmp(dh->q, q))
+    /* check that the q is correct */
+    if (q != NULL) {
```

Comment:
If it was created using DH_new_by_nid() then that is true, otherwise it may not be. For example TLS sends only a p & g to a client. The client then either fails because there is no q or it figures out its a safe prime group (and maybe sets the q itself).